src/transport/TCP.h

```diff
@@ -0,0 +1,235 @@
+     * Initialize a TCP transport on a given port.
+     *
+     * @param params        TCP configuration parameters for this transport
+     *
+     * @details
+     *   Generally send and receive ports should be the same and equal to CHIP_PORT.
+     *   The class allows separate definitions to allow local execution of several
+     *   Nodes.
+     */
+    CHIP_ERROR Init(TcpListenParameters & params);
+
+    CHIP_ERROR SendMessage(const MessageHeader & header, const PeerAddress & address, System::PacketBuffer * msgBuf) override;
+
+    void Disconnect(const PeerAddress & address) override;
+
+    bool CanSendToPeer(const PeerAddress & address) override
+    {
+        return (mState == State::kInitialized) && (address.GetTransportType() == Type::kTcp) &&
+            (address.GetIPAddress().Type() == mEndpointType);
+    }
+
+    /**
+     * Helper method to determine if IO processing is still required for a TCP transport
+     * before everything is cleaned up (socket closing is async, so after calling 'Close' on
+     * the transport, some time may be needed to actually be able to close.)
+     */
+    bool HasActiveConnections() const;
+
+    /**
+     * Close all active connetions
+     */
+    void CloseActiveConnections();
+
+private:
+    /**
+     * Find an active connection to the given peer or return nullptr if
+     * no active connection exists.
+     */
+    Inet::TCPEndPoint * FindActiveConnection(const PeerAddress & addr);
+
+    /**
+     * Sends the specified message once a connection has been established.
+     *
+     * @param addr - what peer to connect to
+     * @param msg - what buffer to send once a connection has been established.
+     */
+    CHIP_ERROR SendAfterConnect(const PeerAddress & addr, System::PacketBuffer * msg);
+
+    /**
+     * Process a single received buffer from the specified peer address.
+     *
+     * @param endPoint the source end point from which the data comes from
+     * @param peerAddress the peer the data is coming from
+     * @param buffer the actual data
+     */
+    CHIP_ERROR ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const PeerAddress & peerAddress, System::PacketBuffer * buffer);
+
+    /**
+     * Process a single message of the specified size from a buffer.
+     *
+     * @param peerAddress the peer the data is coming from
+     * @param buffer the buffer containing the message
+     * @param messageSize how much of the head contains the actual message
+     */
+    CHIP_ERROR ProcessSingleMessageFromBufferHead(const PeerAddress & peerAddress, System::PacketBuffer * buffer,
+                                                  uint16_t messageSize);
+
+    // TCP message receive handler.
+    static void OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuffer * buffer);
+
+    // Called when a connection has been completed
+    static void OnConnectionComplete(Inet::TCPEndPoint * endPoint, INET_ERROR err);
+
+    // Called when a connection has been closed
```

Comment:
Might be worth documenting how this differs from `OnPeerClosed`...

Reply:
I updated this mentioning this originates from TCPEndPoint. 
That class has two callbacks. I am actually unsure why - peerclosed is on read 0 (peer closed) however I am unclear how this one is called (OnClose and OnAbort actually null-out the callbacks with the comment 'application called close explicitly')